core/src/main/java/org/elasticsearch/index/IndexService.java

```diff
@@ -592,6 +592,11 @@ public synchronized void updateMetaData(final IndexMetaData metadata) {
                 rescheduleFsyncTask(durability);
             }
         }
+
+        // update primary terms
+        for (final IndexShard shard : this.shards.values()) {
+            shard.updatePrimaryTerm(metadata.primaryTerm(shard.shardId().id()));
```

Comment:
Agreed this might be a good way to go, but at least now, `org.elasticsearch.index.IndexSettings#updateIndexMetaData` returns true if the setting part of the metadata has changed and we use that to optimize a call to IndexShard. We could capture this information and push it down to index shard (so it can call the engine if settings have changed and do other stuff otherwise) but I rather burden this PR with this.